blog post: Julia 1.7 Highlights (#1419)
JuliaLang/www.julialang.org: blog/2021/11/julia-1.7-highlights.md

default random number generator.
We knew we might want to re-evaluate that choice at some point, but it didn't seem
particularly urgent until Chet Hega [pointed out](https://github.com/JuliaLang/julia/pull/34852)
that by changing algorithms we could not only get a significant speed-up, but also
make random number streams *reproducible in multi-threaded programs*.

Mersenne Twister is famous for having an exceptionally long period, but that requires
a correspondingly large state, and is not truly necessary for any practical application.
We also used thread-local RNG states for thread safety, which has a bit of overhead and made
random streams task-schedule-dependent.
Chet's proposal took advantage of the much smaller state of the [Xoshiro256 family](https://prng.di.unimi.it/)
of RNGs to put a state in every *task*, and fork it on each task creation.
That makes random numbers depend only on the task spawning structure of a program, and not
on the parallel execution schedule.
It took a bit of debate for all of us to be comfortable spending precious bytes of Task
objects this way, but we are huge fans of reproducibility and so the proposal was eventually
adopted.

> ## New atomic support
> *Jameson Nash*
> 
> Atomic accessors support have been expanded to provide more efficient build-blocks for working with threads.
> A summary of the new functionalities was presented at JuliaCon and can be found [here](https://www.youtube.com/watch?v=2rBv6sV4Xts)

Reviewer: ```suggestion
## New Threading Capabilities
*Jameson Nash*

The last few releases have been big ones for multithreading-demanding programs in Julia. This release is no exception: we've addressed numerous race conditions in the runtime, chased down synchronization mistakes, refined support for scheduling workloads on multiple threads, made the default random number generator more thread-friendly, and added atomics as a primitive language feature!

Support for atomic access to mutable struct fields has been added to provide more efficient build-blocks for working with threads via a set of `@atomic` macros, and their corresponding generic functions. A summary of the new functionalities was presented at JuliaCon and can be found [on youtube](https://www.youtube.com/watch?v=2rBv6sV4Xts). This work is based on the "Atomics Manifesto" that described how we envisioned threading to be a key part of the language's future, links to which can be found in the [Julia manual on multi-threading with per-field atomics](https://docs.julialang.org/en/v1.7/manual/multi-threading/#man-atomics)

We are very excited about this. We've already seen some packages in the ecosystem building about this foundation to provide thread-safe abstractions:

* [JuliaConcurrent/ConcurrentCollections.jl](https://github.com/JuliaConcurrent/ConcurrentCollections.jl): queues and dictionaries for shared state.
* [JuliaActors/Actors.jl](https://github.com/JuliaActors/Actors.jl/blob/master/README.md): patterns for making concurrency easy to understand and reason about.
* [tkf/ThreadsX.jl](https://github.com/tkf/ThreadsX.jl): Parallelized versions of some Base functions.
* [JuliaFolds/Transducers.jl](https://juliafolds.github.io/Transducers.jl/dev/parallelism/#overview-parallel): thread-based (foldxt) and process-based (foldxd) parallelisms with the same composable API; i.e. transducers.
* [JuliaFolds/FLoops.jl](https://github.com/JuliaFolds/FLoops.jl/blob/master/README.md): generates a fast generic sequential and parallel iteration over complex collections.
```

Instead of using a thread pool to speed up closing files, we decided to take a different route.

> Julia comes bundled with [p7zip](http://p7zip.sourceforge.net/) and together with the standard library [`Tar.jl`](https://github.com/JuliaIO/Tar.jl) it is possible to directly read the compressed tarball into memory without materializing any files at all. By doing so the problem of materializing files is obliviated which significantly improves the performance of the registry on Windows, NFS and other distributed file systems, such as those typically employed in HPC systems.

Reviewer 2: ```suggestion
Julia comes bundled with [p7zip](http://p7zip.sourceforge.net/) and together with the standard library [`Tar.jl`](https://github.com/JuliaIO/Tar.jl) it is possible to directly read the compressed tarball into memory without materializing any files at all. By doing so the problem of materializing files is obliviated which significantly improves the performance of the registry on Windows, NFS and other distributed file systems typically used in HPC systems.
```

Not a native speaker, but isn't the word employed reserved for people?

Reviewer 3: no. this is a relatively common use. we could still change it if we think it is unclear to non native speakers though.

Author: Just googling how it is described:

> Employ is a formal word when it is used to talk about such things as methods or tools. You usually say that a method or tool is used.

So changing it to used makes sense to me.

Reviewer 3: If we don't like "such as those" we should perhaps change to "like" instead of nothing since distributed file systems aren't only used in HPC systems.

Reviewer 2: the proposed change even drops "like"

+++
mintoclevel = 2
maxtoclevel = 2
title = "Julia 1.7 Highlights"

> +++

Reviewer 4: ```suggestion
authors = "..."
published = "30 November 2021"
rss_pubdate = Date(2021, 11, 30)
rss = """Some highlights of the Julia 1.7 release."""
+++
```

After 4 betas and 3 release candidates, Julia version 1.7 has finally been released. We would like to thank all the contributors to this release and all the testers that helped with finding regressions and issues in the pre-releases. Without you, this release would not have been possible.

The full list of changes can be found in the [NEWS file](https://github.com/JuliaLang/julia/blob/release-1.7/NEWS.md), but here we'll give a more in-depth overview of some of the release highlights.

\toc

## New RNG (reproducible RNG in tasks)
*Jeff Bezanson*

From its very first release, Julia used the popular Mersenne Twister algorithm as its
default random number generator.
We knew we might want to re-evaluate that choice at some point, but it didn't seem
particularly urgent until Chet Hega [pointed out](https://github.com/JuliaLang/julia/pull/34852)
that by changing algorithms we could not only get a significant speed-up, but also

> make random number streams *reproducible in multi-threaded programs*.

Reviewer 5: Looks fine to me here: https://julialang.netlify.app/previews/pr1419/blog/2021/11/julia-1.7-highlights/#new_rng_reproducible_rng_in_tasks

a correspondingly large state, and is not truly necessary for any practical application.
We also used thread-local RNG states for thread safety, which has a bit of overhead and made
random streams task-schedule-dependent.
Chet's proposal took advantage of the much smaller state of the [Xoshiro256 family](https://prng.di.unimi.it/)
of RNGs to put a state in every *task*, and fork it on each task creation.
That makes random numbers depend only on the task spawning structure of a program, and not
on the parallel execution schedule.
It took a bit of debate for all of us to be comfortable spending precious bytes of Task
objects this way, but we are huge fans of reproducibility and so the proposal was eventually
adopted.

## New Threading Capabilities
*Jameson Nash*

The last few releases have been big ones for multithreading-demanding programs in Julia. This release is no exception: we've addressed numerous race conditions in the runtime, chased down synchronization mistakes, refined support for scheduling workloads on multiple threads, made the default random number generator more thread-friendly, and added atomics as a primitive language feature!

> Support for atomic access to mutable struct fields has been added to provide more efficient build-blocks for working with threads via a set of `@atomic` macros, and their corresponding generic functions. A summary of the new functionalities was presented at JuliaCon and can be found [on youtube](https://www.youtube.com/watch?v=2rBv6sV4Xts). This work is based on the "Atomics Manifesto" that described how we envisioned threading to be a key part of the language's future, links to which can be found in the [Julia manual on multi-threading with per-field atomics](https://docs.julialang.org/en/v1.7/manual/multi-threading/#man-atomics)

Reviewer: v1 doesn't have this section, but the released julia should

> Support for atomic access to mutable struct fields has been added to provide more efficient build-blocks for working with threads via a set of `@atomic` macros, and their corresponding generic functions. A summary of the new functionalities was presented at JuliaCon and can be found [on youtube](https://www.youtube.com/watch?v=2rBv6sV4Xts). This work is based on the "Atomics Manifesto" that described how we envisioned threading to be a key part of the language's future, links to which can be found in the [Julia manual on multi-threading with per-field atomics](https://docs.julialang.org/en/v1.7/manual/multi-threading/#man-atomics)

Reviewer 6: the above link works now because `v1` docs is out

+++
mintoclevel = 2
maxtoclevel = 2
title = "Julia 1.7 Highlights"

> authors = "..."

Reviewer 7: I presume this needs to be filled?
```suggestion
authors = "Jeff Bezanson, Jameson Nash, Ian Butterworth, Kristoffer Carlsson, Shuhei Kadowaki, Elliot Saba, Viral B Shah, Mosè Giordano, Simeon Schaub, Nicholas Bauer, Keno Fischer"
```
Or the `"..."` does something special?

Reviewer 6: https://github.com/JuliaLang/www.julialang.org/blob/ca2f054294368ac0964b91cf353db61572eb51b5/previews/PR1419/blog/2021/11/julia-1.7-highlights/index.html#L1

it does expand automatically it seems

Reviewer 4: Yea it was just a placeholder. I don't actually know who have authored the post.

Reviewer 7: I went with the authors of the sections :stuck_out_tongue:
